2020-02-Spring/4_graphs/README.md

```diff
@@ -119,21 +175,170 @@ Go to [Top](#top)
+  }
+}
+```
+###### Source code
+<details>
+<summary>Click to see C++ source code</summary>
+
+```c++
+class Solution{
+public:
+  vector<bool> Maint;
+  vector<bool> PowPlants;
+  void DFS(Graph &G, int Cur, vector<bool> &ActiveStas, vector<bool> &Visited) {
+    if (Maint[Cur]) return;
+    Visited[Cur] = true;
+    ActiveStas[Cur] = true;
+    for (int i=0; i<G.N; i++) {
+      if (i!=Cur && !Visited[i] && G.Matrix[Cur][i]) {
+        DFS(G, i, ActiveStas, Visited);
+      }
+    }
+  }
+  void Solve(Graph &G) {
+    // init
+    Maint = vector<bool>(G.N);
+    PowPlants = vector<bool>(G.N);
+    for (int i=0; i<G.K; i++)
+      Maint[G.Maint[i]] = true;
+    for (int i=0; i<G.M; i++)
+      PowPlants[G.PowPlants[i]] = true;
+
+    // assuming all stations are off except active power plants
+    vector<bool> ActiveStas(G.N);
+    vector<bool> Visited(G.N);
+    for (int i=0; i<G.N; i++) {
+      ActiveStas[i] = false;
+      if (PowPlants[i] && !Maint[i])
+        ActiveStas[i] = true;
+      Visited[i] = false;
+    }
+
+    // using DFS to mark stations that connected to
+    // known active stations
+    for (int i=0; i<G.M; i++) {
+      if (ActiveStas[G.PowPlants[i]])
+        DFS(G, G.PowPlants[i], ActiveStas, Visited);
+    }
+
+    // print out offline stations at the end
+    for (int i=0;i<G.N; i++) {
+      if (ActiveStas[i]) continue;
+      cout << i << " ";
+    }
+    cout << endl;
+  }
+};
+
+```
+</details>
+
+<details>
+<summary>Click to see Java source code</summary>
+
+```Java
+static class Graph {
+  public Graph() { }
+  public int N,M,K;
+  public int Matrix[][];
+  public int PowPlants[];
+  public int Maint[];
+}
+static boolean Maint[];
+static boolean PowPlants[];
+private static void DFS(Graph G, int Cur, boolean ActiveStas[], boolean Visited[]) {
+  if (Maint[Cur]) return;
+  Visited[Cur] = true;
+  ActiveStas[Cur] = true;
+  for (int i=0; i<G.N; i++) {
+    if (i!=Cur && !Visited[i] && G.Matrix[Cur][i] == 1) {
+      DFS(G, i, ActiveStas, Visited);
+    }
+  }
+}
+public static void Solve(Graph G) {
+  // init
+  Maint = new boolean[G.N];
+  PowPlants = new boolean[G.N];
+  for (int i=0; i<G.K; i++)
+    Maint[G.Maint[i]] = true;
+  for (int i=0; i<G.M; i++)
+    PowPlants[G.PowPlants[i]] = true;
+
+  // assuming all stations are off except active power plants
+  boolean ActiveStas[] = new boolean[G.N];
+  boolean Visited[] = new boolean[G.N];
+  for (int i=0; i<G.N; i++) {
+    ActiveStas[i] = false;
+    if (PowPlants[i] && !Maint[i]) ActiveStas[i] = true;
+      Visited[i] = false;
+  }
+  // using DFS to mark stations that connected to
+  // known active stations
+  for (int i=0; i<G.M; i++) {
+    if (ActiveStas[G.PowPlants[i]])
+      DFS(G, G.PowPlants[i], ActiveStas, Visited);
+  }
+
+  // print out offline stations at the end
+  for (int i=0;i<G.N; i++) {
+    if (ActiveStas[i]) continue;
+    System.out.print(i +" ");
+  }
+  System.out.println();
+}
+```
+
+</details>
+
+###### Complexity:
```

Comment:
Is this space or time complexity? It needs to be labeled. Add the other if you know it/can find it,